### PR TITLE
[tt-crank] fuse cross entropy onto the ttml cross_entropy_fw/bw composites

### PR DESCRIPTION
**Pull request by agobeljicTT**
### Ticket
Tracking: tenstorrent/tt-blacksmith#685

### Problem description
`F.cross_entropy` decomposes to `_log_softmax` + `nll_loss_forward` (backward: `nll_loss_backward` + `_log_softmax_backward_data`): four passes over the `[rows x C]` logits per step, two of them f32. ttml has one kernel each way, `cross_entropy_fw` and `cross_entropy_bw`. With Llama's 128k vocabulary this is the largest non-matmul cost in the step.

### What's changed
- `tt_crank::cross_entropy_fw/bw` custom ops as the graph form of the kernels, lowered to `ttcore.composite`s: `OPT_LEVEL 1` promotes to `ttnn.cross_entropy_fw/bw`, `OPT_LEVEL 0` inlines the plain-TTIR decomposition (`build_cross_entropy_fw/bw`, engine).
- `_fuse_cross_entropy` rewrites the aten pattern onto them in the joint graph (partition hook), so aot saves the logits and the row mask instead of the log-probabilities. Fused when: bf16 `[rows x C]` logits, class dim last, no class weights, mean/sum reduction. `ignore_index` rows are selected out around the kernels (`where`, not a multiply, so a non-finite loss in an ignored row cannot leak) and the mean backward clamps the row count to >= 1 (all-ignored batch gives zero gradients like torch), same as the aten lowerings from #9339. Everything else stays on those lowerings.

### Dependencies
- Stacked on #9347 (`ModuleBuilder::create_composite`). Retarget to `main` once it merges.
- #9350 (GQA peel) also defines `_tt_partition`; whichever lands second folds the two rewrites into one hook.
- #9345 (sharded CPU fallback) is not a blocker: this PR is single-chip and never hits the fallback.

### Checklist
- [x] New/Existing tests provide coverage for changes: `tests/python/test_cross_entropy.py`, fused and aten paths at OPT 1 and OPT 0 vs CPU (mean/sum, ignore_index, off-tile shapes, 8k classes, `none`/f32 fallbacks), promotion check on the TTIR/TTNN dumps, edge cases (large logits, all-ignored batch, non-finite logits in ignored rows, rank-1 rejection). OPT 1 cases `xfail(run=False)` under ttsim.

### CODE REVIEW

File: tt-crank/python/tt_crank/torch/_compile.py [hunk 1]
```diff
@@ -970,6 +971,180 @@ def _(
...
+# decompositions. ttml has one kernel each way: cross_entropy_fw (per-row loss from logits and integer
+# targets) and cross_entropy_bw ((softmax - onehot) * grad). The custom ops are their graph form.
+# `aten.cross_entropy_loss` is CompositeImplicitAutograd, so it is already the four aten ops by the time
+# aot traces; the dynamo graph still has the single `F.cross_entropy` call, and _rewrite_cross_entropy
+# swaps that node onto _TTCrossEntropy (an autograd.Function around the two ops) before aot. The aten
+# lowerings below stay as the fallback for whatever the kernels do not take.
+@torch.library.custom_op("tt_crank::cross_entropy_fw", mutates_args=())
+def _cross_entropy_fw(logits: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+    log_probs = torch.log_softmax(logits.float(), dim=-1)
+    return -log_probs.gather(1, target.long().unsqueeze(1)).squeeze(1).to(logits.dtype)
```
Comment (agobeljicTT) [inline tt-crank/python/tt_crank/torch/_compile.py]: @pilkicTT @acolicTT perhaps should have a dedicated file for our custom ops...

File: tt-crank/python/tt_crank/torch/_compile.py [hunk 2]
```diff
@@ -1570,6 +1589,7 @@ def tt_backend(
     options: dict[CompileOption, str | int | bool] | None = None,
 ):
     """Top-level dynamo backend. Delegates to aot_module_simplified."""
+    rewrite_cross_entropy(gm)
```
Comment (agobeljicTT) [inline tt-crank/python/tt_crank/torch/_compile.py:1592]: good place to have a wrapper `rewrite_custom_ops` and group all new thing, but we can let it be like this for now

File: tt-crank/python/tt_crank/torch/_distributed.py [hunk 3]
```diff
@@ -153,16 +153,22 @@ def reduce_scatter_tensor_coalesced(self, output_tensors, input_tensors, opts):
...
+        average = opts.reduceOp == dist.ReduceOp.AVG
+        if not average and opts.reduceOp != dist.ReduceOp.SUM:
             raise NotImplementedError(
-                f"tt allreduce: only ReduceOp.SUM is supported, got {opts.reduceOp}"
+                f"tt allreduce: only ReduceOp.SUM and ReduceOp.AVG are supported, got {opts.reduceOp}"
             )
         for t in tensors:
             _native.allreduce_into(t, self.cluster_axis)
+            if average:
+                t.copy_(t.div(self.size()))
```
Comment (agobeljicTT) [inline tt-crank/python/tt_crank/torch/_distributed.py:171]: @pilkicTT @acolicTT I know this deserves a dedicated PR but I am really tired of splitting stuff xd and this is a micro change and can be inlined here
